svj_ntuple_processing/__init__.py

```diff
@@ -588,170 +577,163 @@ def veto_HEM(eta,phi,pt):
     return hemveto
 
-def cr_filter_preselection(array):
-    """
-    Preselection for the control region.
-
-    w.r.t. `filter_preselection`, does not include:
-    - leading ak8 jet pt > 500
-    - rtx > 1.1
-
-    but includes:
-    - at least 2 ak4 jets
-    - both ak4 jets < 2.4 eta
-    """
-    copy = array.copy()
-    a = copy.array
-    cutflow = copy.cutflow
-
-    # AK8Jet.pT>500
-    a = a[ak.count(a['JetsAK8.fCoordinates.fPt'], axis=-1)>=1] # At least one jet
-    cutflow['ak8_njets>0'] = len(a)
-
-    # Triggers
-    trigger_indices = np.array([copy.trigger_branch.index(t) for t in copy.triggers])
-    if len(a):
-        trigger_decisions = a['TriggerPass'].to_numpy()[:,trigger_indices]
-        a = a[(trigger_decisions == 1).any(axis=-1)]
-    cutflow['triggers'] = len(a)
-
+def select_jet_topology(arrays: Arrays):
     # At least 2 AK15 jets
-    a = a[ak.count(a['JetsAK15.fCoordinates.fPt'], axis=-1) >= 2]
-    cutflow['n_ak15jets>=2'] = len(a)
-
-    # At least 2 AK4 jets --> deadcells study
-    a = a[ak.count(a['Jets.fCoordinates.fPt'], axis=-1) >= 2]
-    cutflow['n_ak4jets>=2'] = len(a)
-
+    arrays.array = arrays.array[ak.count(arrays.array['JetsAK15.fCoordinates.fPt'], axis=-1) >= 2]
+    arrays.cutflow['n_ak15jets>=2'] = len(arrays.array)
+    # jetid for AK15 jets
+    arrays.array = arrays.array[arrays.array['JetsAK15_ID'][:,1]>0]
+    arrays.cutflow['ak15jets_id'] = len(arrays.array)
     # subleading eta < 2.4 eta
-    a = a[np.abs(a['JetsAK15.fCoordinates.fEta'][:,1])<2.4]
-    cutflow['ak15j2_eta<2.4'] = len(a)
-
-    # leading and subleading ak4 eta < 2.4 --> deadcells study
-    a = a[np.abs(a['Jets.fCoordinates.fEta'][:,0])<2.4]
-    a = a[np.abs(a['Jets.fCoordinates.fEta'][:,1])<2.4]
-    cutflow['ak4j1j2_eta<2.4'] = len(a)
-
-    # lepton vetoes
-    a = a[(a['NMuons']==0) & (a['NElectrons']==0)]
-    cutflow['nleptons=0'] = len(a)
-
-    # MET filters
-    for b in met_filters:
-        a = a[a[b]!=0] # Pass events if not 0, is that correct?
-    cutflow['metfilter'] = len(a)
-    cutflow['preselection'] = len(a)
-
-    copy.array = a
-    logger.debug('cutflow:\n%s', pprint.pformat(copy.cutflow))
-    return copy
-
+    arrays.array = arrays.array[np.abs(arrays.array['JetsAK15.fCoordinates.fEta'][:,1])<2.4]
+    arrays.cutflow['subl_eta<2.4'] = len(arrays.array)
+    return arrays
 
-def filter_preselection(array, single_muon_cr=False):
-    """Apply the preselection on the array.
-    cuts now in cutflowtable order by default.
-    should be run after the filter_stitch(array) function (?)
 
-    Args:
-        single_muon_cr (bool): If true, *selects* a muon instead of applying the lepton
-            veto. Disables the AK8Jet.pT cut and the triggers, since this mode is used
-            to study the trigger efficiencies.
-    """
-    copy = array.copy()
-    a = copy.array
-    cutflow = copy.cutflow
+def select_trigger(arrays:Arrays):
+    a = arrays.array
```

Comment:
can we avoid the duplication of these assignment operations:
```python
a = arrays.array
...
arrays.array = a
```
in each function by just passing `arrays.array` and have the EventSelector class deal with the assignments?

Reply:
I guess we would also need to pass and return `cutflow`. but that still seems slightly preferable to me.

Reply:
Right, I kept it like this mainly to try and keep the inferface consistent, but if the user-facing function is just filter_preselection and similar, I can just wrap around this.